Subject: [PATCH] Allow configured platform certificate on SDK37

Keep the current scan ABI and Samsung platform fallback. Match only the
configured certificate after existing signing-details verification.

--- a/smali_classes2/com/android/server/pm/InstallPackageHelper.smali
+++ b/smali_classes2/com/android/server/pm/InstallPackageHelper.smali
@@ -16,6 +16,8 @@
 
 .field public final mContext:Landroid/content/Context;
 
+.field public final mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+
 .field public final mDeletePackageHelper:Lcom/android/server/pm/DeletePackageHelper;
 
 .field public final mIncrementalManager:Landroid/os/incremental/IncrementalManager;
@@ -172,6 +174,18 @@
     check-cast p1, Lcom/android/server/pm/UpdateOwnershipHelper;
 
     iput-object p1, p0, Lcom/android/server/pm/InstallPackageHelper;->mUpdateOwnershipHelper:Lcom/android/server/pm/UpdateOwnershipHelper;
+
+    const-string p1, "CONFIG_CUSTOM_PLATFORM_SIGNATURE"
+
+    filled-new-array {p1}, [Ljava/lang/String;
+
+    move-result-object p1
+
+    invoke-static {p1}, Lcom/android/server/pm/InstallPackageHelper;->createSignatures([Ljava/lang/String;)[Landroid/content/pm/Signature;
+
+    move-result-object p1
+
+    iput-object p1, p0, Lcom/android/server/pm/InstallPackageHelper;->mCustomPlatformSignatures:[Landroid/content/pm/Signature;
 
     new-instance p1, Landroid/util/ArraySet;
 
@@ -23180,6 +23194,8 @@
 
     invoke-static {v1, v10, v2, v9}, Lcom/android/server/pm/ScanPackageUtils;->applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z)V
 
+    invoke-virtual {v15, v1}, Lcom/android/server/pm/InstallPackageHelper;->applyCustomPlatformSignaturePolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;)V
+
     iget-object v2, v15, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
 
     iget-object v3, v2, Lcom/android/server/pm/PackageManagerService;->mInjector:Lcom/android/server/pm/PackageManagerServiceInjector;
@@ -23766,6 +23782,8 @@
     move/from16 v4, v16
 
     invoke-static/range {v0 .. v5}, Lcom/android/server/pm/ScanPackageUtils;->collectCertificatesLI(Lcom/android/server/pm/PackageSetting;Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/Settings$VersionInfo;ZZZ)V
+
+    invoke-virtual {v15, v1}, Lcom/android/server/pm/InstallPackageHelper;->setCustomPlatformSigningDetails(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;)V
     :try_end_10
     .catch Lcom/android/server/pm/PackageManagerException; {:try_start_10 .. :try_end_10} :catch_1
     .catchall {:try_start_10 .. :try_end_10} :catchall_4
@@ -24747,6 +24765,8 @@
 
     invoke-static {v1, v10, v2, v5}, Lcom/android/server/pm/ScanPackageUtils;->applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z)V
 
+    invoke-virtual {v15, v1}, Lcom/android/server/pm/InstallPackageHelper;->applyCustomPlatformSignaturePolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;)V
+
     iget-object v2, v15, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
 
     iget-object v2, v2, Lcom/android/server/pm/PackageManagerService;->mLock:Lcom/android/server/pm/PackageManagerTracedLock;
@@ -24887,3 +24907,64 @@
 
     throw v0
 .end method
+
+
+.method public static createSignatures([Ljava/lang/String;)[Landroid/content/pm/Signature;
+    .locals 5
+
+    array-length v0, p0
+    new-array v1, v0, [Landroid/content/pm/Signature;
+    const/4 v2, 0x0
+
+    :next_signature
+    if-ge v2, v0, :done
+    new-instance v3, Landroid/content/pm/Signature;
+    aget-object v4, p0, v2
+    invoke-direct {v3, v4}, Landroid/content/pm/Signature;-><init>(Ljava/lang/String;)V
+    aput-object v3, v1, v2
+    add-int/lit8 v2, v2, 0x1
+    goto :next_signature
+
+    :done
+    return-object v1
+.end method
+
+.method public final applyCustomPlatformSignaturePolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;)V
+    .locals 2
+
+    iget-object v0, p0, Lcom/android/server/pm/InstallPackageHelper;->mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+    invoke-interface {p1}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->getSigningDetails()Landroid/content/pm/SigningDetails;
+    move-result-object v1
+    invoke-virtual {v1}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
+    move-result-object v1
+    invoke-static {v0, v1}, Lcom/android/server/pm/PackageManagerServiceUtils;->compareSignatureArrays([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)I
+    move-result v0
+    if-nez v0, :done
+    const/4 v0, 0x1
+    invoke-interface {p1, v0}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->setSignedWithPlatformKey(Z)Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
+
+    :done
+    return-void
+.end method
+
+.method public final setCustomPlatformSigningDetails(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;)V
+    .locals 2
+
+    iget-object v0, p0, Lcom/android/server/pm/InstallPackageHelper;->mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+    invoke-interface {p1}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->getSigningDetails()Landroid/content/pm/SigningDetails;
+    move-result-object v1
+    invoke-virtual {v1}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
+    move-result-object v1
+    invoke-static {v0, v1}, Lcom/android/server/pm/PackageManagerServiceUtils;->compareSignatureArrays([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)I
+    move-result v0
+    if-nez v0, :done
+    iget-object v0, p0, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
+    iget-object v0, v0, Lcom/android/server/pm/PackageManagerService;->mPlatformPackage:Lcom/android/server/pm/pkg/AndroidPackage;
+    if-eqz v0, :done
+    invoke-interface {v0}, Lcom/android/server/pm/pkg/AndroidPackage;->getSigningDetails()Landroid/content/pm/SigningDetails;
+    move-result-object v0
+    invoke-interface {p1, v0}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->setSigningDetails(Landroid/content/pm/SigningDetails;)Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
+
+    :done
+    return-void
+.end method
